portfolio/src/main/webapp/script.js

```diff
@@ -102,20 +102,119 @@ function getComments() {
+    this.keyedJSON = this.keyedJSONBackup;
+
+    this.naJSON = this.filterJsonByContinent("NA");
+    this.saJSON = this.filterJsonByContinent("SA");
+    this.euJSON = this.filterJsonByContinent("EU");
+    this.afJSON = this.filterJsonByContinent("AF");
+    this.asJSON = this.filterJsonByContinent("AS");
+    this.ocJSON = this.filterJsonByContinent("OC");
+
+    this.naBool = true;
+    this.saBool = true;
+    this.euBool = true;
+    this.afBool = true;
+    this.asBool = true;
+    this.ocBool = true;
   }
 
   getRandCountryCode() {
     return Object.keys(this.keyedJSON)[Math.floor(Math.random() * Object.keys(this.keyedJSON).length)];
   }
 
   removeCountry(countryCode) {
+
+    const continentStr = this.keyedJSON[countryCode].continent;
+
     delete this.keyedJSON[countryCode];
+    delete this.keyedJSONBackup[countryCode];
+
+    if (continentStr === "NA") {
+      delete this.naJSON[countryCode];
+    }
+    else if (continentStr === "SA") {
+      delete this.saJSON[countryCode];
+    }
+    else if (continentStr === "EU") {
+      delete this.euJSON[countryCode];
+    }
+    else if (continentStr === "AF") {
+      delete this.afJSON[countryCode];
+    }
+    else if (continentStr === "AS") {
+      delete this.asJSON[countryCode];
+    }
+    else {
+      delete this.ocJSON[countryCode];
+    }
   }
 
   empty() {
     return Object.keys(this.keyedJSON).length === 0;
   }
+
+  filterJsonByContinent(continentStr) {
+
+    const continentJson = JSON.parse(JSON.stringify(this.keyedJSONBackup));
+
+    Object.keys(continentJson).forEach(countryCode => {
+
+      if (!(continentJson[countryCode].continent == continentStr)) {
+        delete continentJson[countryCode];
+      }
+    });
+    return continentJson;
+  }
+
+  remakeJsonByContinent(nanaBool = this.naBool, saBool = this.saBool, euBool = this.euBool, afBool = this.afBool, asBool = this.asBool, ocBool = this.ocBool) {
+    this.naBool = nanaBool;
+    this.saBool = saBool;
+    this.euBool = euBool;
+    this.afBool = afBool;
+    this.asBool = asBool;
+    this.ocBool = ocBool;
+
+    this.keyedJSON = {};
+
+    if (this.naBool && this.saBool && this.euBool && this.afBool && this.asBool && this.ocBool) {
+      this.keyedJSON = this.keyedJSONBackup;
+    }
+    else {
+
+      if (this.naBool) {
+        Object.assign(this.keyedJSON, this.naJSON);
+      }
+
+      if (this.saBool) {
+        Object.assign(this.keyedJSON, this.saJSON);
+      }
+
+      if (this.euBool) {
+        Object.assign(this.keyedJSON, this.euJSON);
+      }
+
+      if (this.afBool) {
+        Object.assign(this.keyedJSON, this.afJSON);
+      }
+
+      if (this.asBool) {
+        Object.assign(this.keyedJSON, this.asJSON);
+      }
+
+      if (this.ocBool) {
+        Object.assign(this.keyedJSON, this.ocJSON);
+      }
+    }
+
+    if (!(!this.naBool && !this.saBool && !this.euBool && !this.afBool && !this.asBool && !this.ocBool) && Object.keys(this.keyedJSON).length > 0) {
```

Comment:
How about using De Morgan's Law to replace the double negation with `(this.naBool || this.saBool || ... || this.ocBool)`?

Reply:
Ok, I didn't see that optimization. 😃